### PR TITLE
initial MsSQL connector [sc-12761]

### PR DESCRIPTION
**Pull request by pk1230h**
### 🤔 Why?
initial the MsSQL connector and make MsSQL extractor as the library for Synapse

### 🤓 What?
- initial MsSQL connector
- share common SQL query code with Synapse and write into MsSQL extractor
-  remove duplicated methods in Synapse and use MsSQL methods instead
- add MsSQL readMe and adjust Synapse readMe

### 🧪 Tested?
Unit tests.
manual test  to deploy with ingestion API to dev environment

### CODE REVIEW

File: metaphor/synapse/README.md [hunk 1]
```diff
@@ -38,7 +38,7 @@ Create a YAML config file based on the following template.
 ```yaml
 tenant_id: <tenant_id>  # The azure directory (tenant) id
 
-workspace_name: <workspace_name> # The Microsoft Synapse workspace name
```
Comment (pk1230h) [inline metaphor/synapse/README.md:41]: change parameter because reuse code from mssql

File: metaphor/synapse/extractor.py [hunk 2]
```diff
@@ -1,74 +1,73 @@
...
 
     def __init__(self, config: SynapseConfig):
-        super().__init__(config, "Synapse metadata crawler", Platform.SYNAPSE)
-        self._client = WorkspaceClient(
-            config.workspace_name, config.username, config.password
+        super().__init__(
+            config, "Synapse metadata crawler", Platform.SYNAPSE, DataPlatform.SYNAPSE
         )
-        self._tenant_id = config.tenant_id
+        self.config = config
```
Comment (elic-eon) [inline metaphor/synapse/extractor.py:34]:
```suggestion
        self._config = config
```

Comment (elic-eon) [reply metaphor/synapse/extractor.py:34]: Align with other class member?

File: tests/mssql/test_mssql_client.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        row.append(table.type)
+        row.append(table.create_time)
+        row.append(column.name)
+        row.append(column.max_length)
+        row.append(column.precision)
+        row.append(column.is_nullable)
+        row.append(column.type)
+        row.append(column.is_unique)
+        row.append(column.is_primary_key)
+        row.append(table.is_external)
```
Comment (elic-eon) [inline tests/mssql/test_mssql_client.py:83]: Although this is testing code, why not use `row = [table.id, table.name, ... ]`?

File: tests/synapse/test_extractor.py [hunk 4]
```diff
@@ -274,25 +260,26 @@ async def test_dedicated_sql_pool_extractor_with_query_log(test_root_dir):
...
+        mock_client.return_value = mock_client_instance
+        mock_query.return_value = [qyeryLogTable]
 
         config = synapse_config
         config.query_log = SynapseQueryLogConfig(lookback_days=1)
         extractor = SynapseExtractor(config)
 
         events = [EventUtil.trim_event(e) for e in await extractor.extract()]
+    # import json
+    # print(json.dumps(events, indent=2))
```
Comment (elic-eon) [inline tests/synapse/test_extractor.py:282]: Drop these line?

File: metaphor/mssql/ReadMe.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+## Setup
+We recommend creating a dedicated MsSQL user with limited permissions for the connector to use
+
+Use the following command to set up the read-only user for MsSQL database
+
+```sql
+-- nagivate to master database
+CREATE LOGIN <username> WITH PASSWORD = '<password>'
+CREATE USER <username> FROM LOGIN <username>
+-- switch to user  MsSQL database
```
Comment (elic-eon) [inline metaphor/mssql/ReadMe.md:14]:
```suggestion
-- switch to user MsSQL database
```

File: metaphor/mssql/extractor.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            for database in client.get_databases():
+                tables = client.get_tables(database.name)
+                if len(tables) == 0:
+                    continue
+                datasets = self._map_tables_to_dataset(
+                    self.config.server_name, database, tables, client
+                )
+                entities.extend(datasets)
+        except Exception as error:
+            logger.exception(f"serverless sqlpool error: {error}")
```
Comment (elic-eon) [inline metaphor/mssql/extractor.py:66]: Change the logging message?

File: metaphor/mssql/extractor.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+        endpoint = f"{self.config.server_name}.database.windows.net"
+        client = MssqlClient(endpoint, self.config.username, self.config.password)
+
+        entities: List[ENTITY_TYPES] = []
+        try:
+            for database in client.get_databases():
+                tables = client.get_tables(database.name)
+                if len(tables) == 0:
+                    continue
```
Comment (elic-eon) [inline metaphor/mssql/extractor.py:60]: Does this test necessary? `_map_tables_to_dataset` should return an empty result if we pass an empty list

File: metaphor/mssql/extractor.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        for fk in client.get_foreign_keys(database.name):
+            if fk.table_id in dataset_map:
+                foreign_key = ForeignKey(
+                    field_path=fk.column_name,
+                    parent=dataset_map[fk.referenced_table_id].logical_id,
+                    parent_field=fk.referenced_column,
+                )
+                dataset_map[fk.table_id].schema.sql_schema.foreign_key.append(
+                    foreign_key
+                )
```
Comment (elic-eon) [inline metaphor/mssql/extractor.py:141]: Possible to move logic related to foreign keys out from `_map_tables_to_dataset`? Therefore, we don't have to pass the client object in.

File: metaphor/mssql/model.py [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    request_id: str
+    session_id: Optional[str]
+    sql_query: str
+    login_name: str
+    start_time: datetime
+    end_time: datetime
+    # in milliseconds
+    duration: int
+    # in MB
+    query_size: Optional[int]
```
Comment (elic-eon) [inline metaphor/mssql/model.py:60]: Will `duration_in_ms` and `query_size_in_mb` be better here?

File: metaphor/mssql/model.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
+from datetime import datetime
+from typing import Dict, Optional
+
+from pydantic import BaseModel
+
+
+class MssqlConnectConfig(BaseModel):
```
Comment (elic-eon) [inline metaphor/mssql/model.py:7]: Is `MssqlClientConfig` a better name here? We use this class in the `MssqlClient`, right?

File: metaphor/synapse/README.md [hunk 11]
```diff
@@ -38,7 +38,7 @@ Create a YAML config file based on the following template.
...
+server_name: <workspace_name> # The Microsoft Synapse workspace name
```
Comment (elic-eon) [inline metaphor/synapse/README.md:41]:
```suggestion
server_name: <workspace_name>  # The Microsoft Synapse workspace name
```

File: metaphor/synapse/README.md [hunk 12]
```diff
@@ -38,7 +38,7 @@ Create a YAML config file based on the following template.
...
 
 username: <username> # The synapse workspace SQL username
```
Comment (elic-eon) [inline metaphor/synapse/README.md:43]:
```suggestion
username: <username>  # The synapse workspace SQL username
```

File: metaphor/mssql/extractor.py (hunk 8 above)
Comment (pk1230h) [reply metaphor/mssql/extractor.py:141]: The reason to do this way is that I need to have completed dataset_map to add foreign_key.  
if not doing this way, the other way  take the dataset map object as return value and use another static method to add foreign. The method will look like:
```
def _get_foreign_key_for_dataset(
     dataset_map,
     database_name,
     client
)
```
is this the better approach?

File: metaphor/mssql/extractor.py (hunk 7 above)
Comment (pk1230h) [reply metaphor/mssql/extractor.py:60]: just fast check if empty don't need to go to the **_map_tables_to_dataset** and below logic, 
The foreign keys logic is not within tables logic and will trigger MsSQL fetch query call to check if foreign key existed. I  think it no need to make the call if we know table is not existed/

File: metaphor/mssql/extractor.py (hunk 8 above)
Comment (elic-eon) [reply metaphor/mssql/extractor.py:141]: Yes, we can return a map and pass into new function. I think two smaller functions would be better.

File: metaphor/mssql/extractor.py (hunk 7 above)
Comment (elic-eon) [reply metaphor/mssql/extractor.py:60]: My point is to reduce the number of indentions. If we can make sure all functions takes `tables` can handle empty list efficiently, would you think it better to drop these two lines?

File: metaphor/mssql/ReadMe.md [hunk 13]
```diff
@@ -0,0 +1,50 @@
+# MsSQL Connector
+
+This connector extracts technical metadata from Azure Synapse workspaces using [System catalog views](https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/catalog-views-transact-sql?view=sql-server-ver16).
```
Comment (elic-eon) [inline metaphor/mssql/ReadMe.md:3]:
```suggestion
This connector extracts technical metadata from a Microsoft SQL Server using [System catalog views](https://learn.microsoft.com/en-us/sql/relational-databases/system-catalog-views/catalog-views-transact-sql?view=sql-server-ver16).
```

File: metaphor/mssql/extractor.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+    def _map_tables_to_dataset(
+        self,
+        sql_server_name: str,
+        database: MssqlDatabase,
+        tables: List[MssqlTable],
+    ):
+        dataset_map: Dict[str, Dataset] = {}
+        if len(tables) == 0:
+            return dataset_map
```
Comment (elic-eon) [inline metaphor/mssql/extractor.py:76]:
```suggestion
        if len(tables) == 0:
            return dataset_map

        dataset_map: Dict[str, Dataset] = {}
```

**Review by elic-eon — COMMENTED** [review]
Otherwise, LGTM

**Pull request merged**